sycl/test/basic_tests/builtins_implicitly_convertible_args.cpp

```diff
@@ -0,0 +1,172 @@
+// RUN: %clangxx -fsycl -fpreview-breaking-changes -fsyntax-only %s -o %t
+// REQUIRES: preview-breaking-changes-supported
+
+// Checks that builtins without template arguments allow for implicit
+// conversions of arguments.
+
+#include <sycl/sycl.hpp>
+
+template <typename T> struct ImplicitlyConvertibleType {
+  operator T() const { return {}; }
+};
+
+#define UNARY_CHECK_INNER(FUNC_NAME, ...)                                      \
+  static_assert(std::is_same_v<                                                \
+                decltype(sycl::FUNC_NAME(                                      \
+                    std::declval<ImplicitlyConvertibleType<__VA_ARGS__>>())),  \
+                decltype(sycl::FUNC_NAME(std::declval<__VA_ARGS__>()))>);
```

Comment:
I'd prefer this approach instead: https://github.com/intel/llvm/blob/cc59cc6648966d0eeb99aebf82d2721131a9121b/sycl/include/sycl/detail/builtins/helper_macros.hpp#L64-L67

and probably the `FOR_EACHN` from that draft for the types as well.

That said, I'm not sure how useful this test really is.

Reply:
> Could you clarify what you are suggesting here?

Having just one `#define CHECK_INNTER(NUM_ARGS, FUNC_NAME, ...)` macro definition so that we could split logic (assert and its condition) from form (number of arguments of a builtin).

Reply:
Based on our discussion offline I did the refactoring you suggested. Let me know if I misused something.